torch/csrc/jit/codegen/cuda/lower_predicate_elimination.cpp

```diff
@@ -241,6 +247,99 @@ class PredicateChcker : public IterVisitor {
+    // This is initial step to gradually remove predicates around
+    //  sharedmem access in suitable situations.
+    // Using an additional variable to track the predicate-on reasons
+    //  when the predicate around shared mem cannot be removed.
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+        if (producer->getMemoryType() == MemoryType::Shared ||
+            consumer->getMemoryType() == MemoryType::Shared) {
+          if (needSharedMemPredicate(producer, consumer)) {
+            return true;
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  // Check for conditions where the predicate cannot be removed
+  //  when either producer or consumer is in shared memory.
+  bool needSharedMemPredicate(TensorView* producer, TensorView* consumer)
+      const {
+    // Indexing is based on consumer leaf ids so check the consumer.
+    auto& parallel_dimension_map = GpuLower::current()->parallelDimensionMap();
+    for (auto id : consumer->domain()->domain()) {
+      if (id->isThreadDim()) {
+        auto ptype = id->getParallelType();
+        // Need to predicate to avoid out of bound access
+        //  because of over-subscribed block size. 
+        if (!parallel_dimension_map.isExact(ptype)) {
+          return true;
+        }
+      }
+
+      // TODO: (Enable in a follow up)
+      //  smem predicate removal with init would break unroll and unswitch,
+      //  eg. as in issue 1133, so disabling this removal pattern for now.
+      if (id->getParallelType() == ParallelType::Unroll ||
+          id->getParallelType() == ParallelType::Unswitch) {
+        return true;
+      }
+
+      // TODO: （Enable in a follow up)
+      //  This cannot yet be removed since smem initialization needs to be
+      //  handled specially, e.g. as in smem_reduce test. Will be able to
+      //  lift this one once the generic pred removal pass with fusion
+      //  traversal is ready.
+      auto consumer_def = consumer->definition();
+      if (consumer_def->isA<ReductionOp>() || consumer_def->isA<WelfordOp>()) {
+        if (producer->getMemoryType() == MemoryType::Shared) {
+          return true;
+        }
+      }
+    }
+  }
+
+  // Utility to find the leaf iterdomains of the given
+  //   tensor view that will be treated as "zero loops"
+  //   in the indexing pass.
+  std::vector<Val*> getZeroLeafIds(const TensorView* tv) const {
+    TORCH_INTERNAL_ASSERT(
+        output->getMemoryType() == MemoryType::Local ||
+            output->getMemoryType() == MemoryType::Shared,
+        "Local or shared memory tensor is assumed: ",
+        output->toString());
+    bool is_shared_mem = tv->getMemoryType() == MemoryType::Shared;
+    std::vector<Val*> zero_leaf_ids;
+    for (const auto i : c10::irange(tv->nDims())) {
+      auto leaf_id = tv->axis(i);
+      if (is_shared_mem && leaf_id->isThread()) {
```

Comment:
If it's parallelized with `BID`, it is a zero loop. Note that `isThread` includes `isBlockDim`.

Reply:
Good catched. Changed to `isThreadDim` here and updated comment. Thanks.